File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-shared-5-4/Account.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Account : ISerialized
    {
        public Account(XmlNode xmlNode)
        {
            XmlNodeList accountIdNodeList = xmlNode.SelectNodes("accountId");
            
            if (accountIdNodeList != null)
            {
                this.accountId_ = new List<AccountId>();
                foreach (XmlNode item in accountIdNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            accountIdIDRef_ = item.Attributes["id"].Value;
                            accountId_.Add(new AccountId(item));
                            IDManager.SetID(accountIdIDRef_, accountId_[accountId_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            accountIdIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        accountId_.Add(new AccountId(item));
                        }
                    }
                    else
                    {
                        accountId_.Add(new AccountId(item));
                    }
                }
            }
            
        
            XmlNode accountNameNode = xmlNode.SelectSingleNode("accountName");
            
            if (accountNameNode != null)
            {
                if (accountNameNode.Attributes["href"] != null || accountNameNode.Attributes["id"] != null) 
                {
                    if (accountNameNode.Attributes["id"] != null) 
                    {
                        accountNameIDRef_ = accountNameNode.Attributes["id"].Value;
                        AccountName ob = new AccountName(accountNameNode);
                        IDManager.SetID(accountNameIDRef_, ob);
                    }
                    else if (accountNameNode.Attributes["href"] != null)
                    {
                        accountNameIDRef_ = accountNameNode.Attributes["href"].Value;
                    }
                    else
                    {
                        accountName_ = new AccountName(accountNameNode);
                    }
                }
                else
                {
                    accountName_ = new AccountName(accountNameNode);
                }
            }
            
        
            XmlNode accountTypeNode = xmlNode.SelectSingleNode("accountType");
            
            if (accountTypeNode != null)
            {
                if (accountTypeNode.Attributes["href"] != null || accountTypeNode.Attributes["id"] != null) 
                {
                    if (accountTypeNode.Attributes["id"] != null) 
                    {
                        accountTypeIDRef_ = accountTypeNode.Attributes["id"].Value;
                        AccountType ob = new AccountType(accountTypeNode);
                        IDManager.SetID(accountTypeIDRef_, ob);
                    }
                    else if (accountTypeNode.Attributes["href"] != null)
                    {
                        accountTypeIDRef_ = accountTypeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        accountType_ = new AccountType(accountTypeNode);
                    }
                }
                else
                {
                    accountType_ = new AccountType(accountTypeNode);
                }
            }
            
        
            XmlNode accountBeneficiaryNode = xmlNode.SelectSingleNode("accountBeneficiary");
            
            if (accountBeneficiaryNode != null)
            {
                if (accountBeneficiaryNode.Attributes["href"] != null || accountBeneficiaryNode.Attributes["id"] != null) 
                {
                    if (accountBeneficiaryNode.Attributes["id"] != null) 
                    {
                        accountBeneficiaryIDRef_ = accountBeneficiaryNode.Attributes["id"].Value;
                        PartyReference ob = new PartyReference(accountBeneficiaryNode);
                        IDManager.SetID(accountBeneficiaryIDRef_, ob);
                    }
                    else if (accountBeneficiaryNode.Attributes["href"] != null)
                    {
                        accountBeneficiaryIDRef_ = accountBeneficiaryNode.Attributes["href"].Value;
                    }
                    else
                    {
                        accountBeneficiary_ = new PartyReference(accountBeneficiaryNode);
                    }
                }
                else
                {
                    accountBeneficiary_ = new PartyReference(accountBeneficiaryNode);
                }
            }
            
        
            XmlNode servicingPartyNode = xmlNode.SelectSingleNode("servicingParty");
            
            if (servicingPartyNode != null)
            {
                if (servicingPartyNode.Attributes["href"] != null || servicingPartyNode.Attributes["id"] != null) 
                {
                    if (servicingPartyNode.Attributes["id"] != null) 
                    {
                        servicingPartyIDRef_ = servicingPartyNode.Attributes["id"].Value;
                        PartyReference ob = new PartyReference(servicingPartyNode);
                        IDManager.SetID(servicingPartyIDRef_, ob);
                    }
                    else if (servicingPartyNode.Attributes["href"] != null)
                    {
                        servicingPartyIDRef_ = servicingPartyNode.Attributes["href"].Value;
                    }
                    else
                    {
                        servicingParty_ = new PartyReference(servicingPartyNode);
                    }
                }
                else
                {
                    servicingParty_ = new PartyReference(servicingPartyNode);
                }
            }
            
        
        }
        
    
        #region AccountId_
        private List<AccountId> accountId_;
        public List<AccountId> AccountId_
        {
            get
            {
                if (this.accountId_ != null)
                {
                    return this.accountId_; 
                }
                else if (this.accountIdIDRef_ != null)
                {
                    return this.accountId_; 
                }
                else
                {
                      return this.accountId_; 
                }
            }
            set
            {
                if (this.accountId_ != value)
                {
                    this.accountId_ = value;
                }
            }
        }
        #endregion
        
        public string accountIdIDRef_ { get; set; }
        #region AccountName_
        private AccountName accountName_;
        public AccountName AccountName_
        {
            get
            {
                if (this.accountName_ != null)
                {
                    return this.accountName_; 
                }
                else if (this.accountNameIDRef_ != null)
                {
                    accountName_ = IDManager.getID(accountNameIDRef_) as AccountName;
                    return this.accountName_; 
                }
                else
                {
                      return this.accountName_; 
                }
            }
            set
            {
                if (this.accountName_ != value)
                {
                    this.accountName_ = value;
                }
            }
        }
        #endregion
        
        public string accountNameIDRef_ { get; set; }
        #region AccountType_
        private AccountType accountType_;
        public AccountType AccountType_
        {
            get
            {
                if (this.accountType_ != null)
                {
                    return this.accountType_; 
                }
                else if (this.accountTypeIDRef_ != null)
                {
                    accountType_ = IDManager.getID(accountTypeIDRef_) as AccountType;
                    return this.accountType_; 
                }
                else
                {
                      return this.accountType_; 
                }
            }
            set
            {
                if (this.accountType_ != value)
                {
                    this.accountType_ = value;
                }
            }
        }
        #endregion
        
        public string accountTypeIDRef_ { get; set; }
        #region AccountBeneficiary_
        private PartyReference accountBeneficiary_;
        public PartyReference AccountBeneficiary_
        {
            get
            {
                if (this.accountBeneficiary_ != null)
                {
                    return this.accountBeneficiary_; 
                }
                else if (this.accountBeneficiaryIDRef_ != null)
                {
                    accountBeneficiary_ = IDManager.getID(accountBeneficiaryIDRef_) as PartyReference;
                    return this.accountBeneficiary_; 
                }
                else
                {
                      return this.accountBeneficiary_; 
                }
            }
            set
            {
                if (this.accountBeneficiary_ != value)
                {
                    this.accountBeneficiary_ = value;
                }
            }
        }
        #endregion
        
        public string accountBeneficiaryIDRef_ { get; set; }
        #region ServicingParty_
        private PartyReference servicingParty_;
        public PartyReference ServicingParty_
        {
            get
            {
                if (this.servicingParty_ != null)
                {
                    return this.servicingParty_; 
                }
                else if (this.servicingPartyIDRef_ != null)
                {
                    servicingParty_ = IDManager.getID(servicingPartyIDRef_) as PartyReference;
                    return this.servicingParty_; 
                }
                else
                {
                      return this.servicingParty_; 
                }
            }
            set
            {
                if (this.servicingParty_ != value)
                {
                    this.servicingParty_ = value;
                }
            }
        }
        #endregion
        
        public string servicingPartyIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-shared-5-4/Account.cpp
// Account.cpp 
#include "Account.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Account::Account(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //accountIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* accountIdNode = xmlNode->FirstChildElement("accountId");

   if(accountIdNode){accountIdIsNull_ = false;}
   else{accountIdIsNull_ = true;}

   if(accountIdNode)
   {
      for(accountIdNode; accountIdNode; accountIdNode = accountIdNode->NextSiblingElement("accountId")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- accountIdNode , address : " << accountIdNode << std::endl;
          #endif
          if(accountIdNode->Attribute("href") || accountIdNode->Attribute("id"))
          {
              if(accountIdNode->Attribute("id"))
              {
                  accountIdIDRef_ = accountIdNode->Attribute("id");
                  accountId_.push_back(boost::shared_ptr<AccountId>(new AccountId(accountIdNode)));
                  accountId_.back()->setID(accountIdIDRef_);
                  IDManager::instance().setID(accountIdIDRef_, accountId_.back());
              }
              else if(accountIdNode->Attribute("href")) { accountIdIDRef_ = accountIdNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { accountId_.push_back(boost::shared_ptr<AccountId>(new AccountId(accountIdNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- accountIdNode , address : " << accountIdNode << std::endl;
       #endif
   }

   //accountNameNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* accountNameNode = xmlNode->FirstChildElement("accountName");

   if(accountNameNode){accountNameIsNull_ = false;}
   else{accountNameIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- accountNameNode , address : " << accountNameNode << std::endl;
   #endif
   if(accountNameNode)
   {
      if(accountNameNode->Attribute("href") || accountNameNode->Attribute("id"))
      {
          if(accountNameNode->Attribute("id"))
          {
             accountNameIDRef_ = accountNameNode->Attribute("id");
             accountName_ = boost::shared_ptr<AccountName>(new AccountName(accountNameNode));
             accountName_->setID(accountNameIDRef_);
             IDManager::instance().setID(accountNameIDRef_,accountName_);
          }
          else if(accountNameNode->Attribute("href")) { accountNameIDRef_ = accountNameNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { accountName_ = boost::shared_ptr<AccountName>(new AccountName(accountNameNode));}
   }

   //accountTypeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* accountTypeNode = xmlNode->FirstChildElement("accountType");

   if(accountTypeNode){accountTypeIsNull_ = false;}
   else{accountTypeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- accountTypeNode , address : " << accountTypeNode << std::endl;
   #endif
   if(accountTypeNode)
   {
      if(accountTypeNode->Attribute("href") || accountTypeNode->Attribute("id"))
      {
          if(accountTypeNode->Attribute("id"))
          {
             accountTypeIDRef_ = accountTypeNode->Attribute("id");
             accountType_ = boost::shared_ptr<AccountType>(new AccountType(accountTypeNode));
             accountType_->setID(accountTypeIDRef_);
             IDManager::instance().setID(accountTypeIDRef_,accountType_);
          }
          else if(accountTypeNode->Attribute("href")) { accountTypeIDRef_ = accountTypeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { accountType_ = boost::shared_ptr<AccountType>(new AccountType(accountTypeNode));}
   }

   //accountBeneficiaryNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* accountBeneficiaryNode = xmlNode->FirstChildElement("accountBeneficiary");

   if(accountBeneficiaryNode){accountBeneficiaryIsNull_ = false;}
   else{accountBeneficiaryIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- accountBeneficiaryNode , address : " << accountBeneficiaryNode << std::endl;
   #endif
   if(accountBeneficiaryNode)
   {
      if(accountBeneficiaryNode->Attribute("href") || accountBeneficiaryNode->Attribute("id"))
      {
          if(accountBeneficiaryNode->Attribute("id"))
          {
             accountBeneficiaryIDRef_ = accountBeneficiaryNode->Attribute("id");
             accountBeneficiary_ = boost::shared_ptr<PartyReference>(new PartyReference(accountBeneficiaryNode));
             accountBeneficiary_->setID(accountBeneficiaryIDRef_);
             IDManager::instance().setID(accountBeneficiaryIDRef_,accountBeneficiary_);
          }
          else if(accountBeneficiaryNode->Attribute("href")) { accountBeneficiaryIDRef_ = accountBeneficiaryNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { accountBeneficiary_ = boost::shared_ptr<PartyReference>(new PartyReference(accountBeneficiaryNode));}
   }

   //servicingPartyNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* servicingPartyNode = xmlNode->FirstChildElement("servicingParty");

   if(servicingPartyNode){servicingPartyIsNull_ = false;}
   else{servicingPartyIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- servicingPartyNode , address : " << servicingPartyNode << std::endl;
   #endif
   if(servicingPartyNode)
   {
      if(servicingPartyNode->Attribute("href") || servicingPartyNode->Attribute("id"))
      {
          if(servicingPartyNode->Attribute("id"))
          {
             servicingPartyIDRef_ = servicingPartyNode->Attribute("id");
             servicingParty_ = boost::shared_ptr<PartyReference>(new PartyReference(servicingPartyNode));
             servicingParty_->setID(servicingPartyIDRef_);
             IDManager::instance().setID(servicingPartyIDRef_,servicingParty_);
          }
          else if(servicingPartyNode->Attribute("href")) { servicingPartyIDRef_ = servicingPartyNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { servicingParty_ = boost::shared_ptr<PartyReference>(new PartyReference(servicingPartyNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
std::vector<boost::shared_ptr<AccountId>> Account::getAccountId()
{
   if(!this->accountIdIsNull_){
        if(accountIdIDRef_ != ""){
             return this->accountId_;
        }else{
             return this->accountId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<AccountId>>();
   }
}
boost::shared_ptr<AccountName> Account::getAccountName()
{
   if(!this->accountNameIsNull_){
        if(accountNameIDRef_ != ""){
             return boost::shared_static_cast<AccountName>(IDManager::instance().getID(accountNameIDRef_));
        }else{
             return this->accountName_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AccountName>();
   }
}
boost::shared_ptr<AccountType> Account::getAccountType()
{
   if(!this->accountTypeIsNull_){
        if(accountTypeIDRef_ != ""){
             return boost::shared_static_cast<AccountType>(IDManager::instance().getID(accountTypeIDRef_));
        }else{
             return this->accountType_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AccountType>();
   }
}
boost::shared_ptr<PartyReference> Account::getAccountBeneficiary()
{
   if(!this->accountBeneficiaryIsNull_){
        if(accountBeneficiaryIDRef_ != ""){
             return boost::shared_static_cast<PartyReference>(IDManager::instance().getID(accountBeneficiaryIDRef_));
        }else{
             return this->accountBeneficiary_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PartyReference>();
   }
}
boost::shared_ptr<PartyReference> Account::getServicingParty()
{
   if(!this->servicingPartyIsNull_){
        if(servicingPartyIDRef_ != ""){
             return boost::shared_static_cast<PartyReference>(IDManager::instance().getID(servicingPartyIDRef_));
        }else{
             return this->servicingParty_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PartyReference>();
   }
}
}

